mixins/labelled-mixin.js

```diff
@@ -0,0 +1,87 @@
+
+export const LabelMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			_label: { type: String, reflect: true }
+		};
+	}
+
+};
+
+export const LabelledMixin = superclass => class extends superclass {
+
+	static get properties() {
+		return {
+			labelledBy: { type: String, reflect: true, attribute: 'labelled-by' },
+			label: { type: String }
+		};
+	}
+
+	updated(changedProperties) {
+		super.updated(changedProperties);
+
+		if (!changedProperties.has('labelledBy')) return;
+
+		if (this._labelObserver) this._labelObserver.disconnect();
+
+		if (this.labelledBy) {
+
+			let labelElem = this.getRootNode().querySelector(`#${this.labelledBy}`);
+			const ancestor = this._getCommonAncestor(labelElem);
+
+			const getLabel = () => {
+				if (labelElem.tagName.includes('-')) return labelElem._label;
+				else return labelElem.textContent;
+			};
+
+			this._labelObserver = new MutationObserver((mutations) => {
+				for (let i = 0; i < mutations.length; i++) {
+
+					for (let j = 0; j < mutations[i].removedNodes.length; j++) {
+						if (mutations[i].removedNodes[j] === labelElem) {
+							labelElem = null;
+							break;
+						}
+					}
+
+					if (!labelElem && mutations[i].addedNodes.length > 0) {
+						labelElem = this.getRootNode().querySelector(`#${this.labelledBy}`);
+						break;
+					}
+
+				}
+
+				if (labelElem) {
+					this.label = getLabel();
+				} else {
+					console.warn('LabelledMixin with specified labelledBy but no such element exists.');
+					this.label = undefined;
+				}
+
+			});
+			this._labelObserver.observe(ancestor, { attributes: true, characterData: true, childList: true, subtree: true });
```

Comment:
Do we know at this point if we could use one mechanism or the other? i.e. do we know here that we're dealing with a web component vs. not? If so, then it might be worth the performance boost to have a different config for each of those two cases.

Also, could `attributeFilter: ['_label']` be used to narrow that down? I think legacy-Edge still needs `attributes: true` [as I discovered here](https://github.com/BrightspaceUI/core/blob/1a9b1ed4818432f54c5e62846e83d8db32c9e262/components/table/table-wrapper.js#L327), but at least we could comment that and then remove it in a month.

Reply:
Yeah that's a good point.  I will update.